Выполнил домашнее задание №5

MAMKUH6OPIIIUK opened the pull request:
Новая работа. 
Чеклист выполненных пунктов задания:
1. Выполнено с H2 для простоты. Для всех запросов используется NamedParameterJdbcOperations
2. Предусмотрены указанные таблицы и дополнительная таблица-связка книг и жанров
3. Сделал с усложнением - у книги может быть много жанров, но всегда один автор. Постарался сделать работу с БД +- оптимальной.
4. Реализован интерфейс на Spring Shell:

-  под каждую сущность - жанры, авторы, книги - своя группа команд в отдельном ShellComponent. Дополнительный компонент для включения консоли.
- реализован набор Component, выполняющих форматирование доменных объектов к строкам. Эти компоненты вызываются из Shell для отображения данных пользователю

5. Скрипты schema.sql и data.sql (data для тестов - отдельный) реализовал, прописал их использование в yml 
6. Написал тесты для всех методов всех дао с использованием JdbcTest. В некоторых тестах контекст пересоздается после выполнения теста, т.к. они могут задевать autoincrement идентификаторы в БД. Тест сервиса работы с книгами реализовал в виде unit test с замоканным dao. 

Включил логирование выполняемых запросов к БД, чтобы при проверке сразу было видно, какие запросы при выполнении команд дёргаются

**Есть сомнения по поводу распределения ответственности:**
1. Transactional развешены в дао. Возможно, стоило делать их в сервисе?
2. Дао жанров имеет метод поиска всех жанров по идентификатору книги. Это используется для оптимального поиска конкретной книги в дао для работы с книгами. Возможно, стоило делать этот запрос прямо в дао для книг? 
3. При создании\обновлении книги в дао передается неполноценный доменный объект - в качестве автора и жанров там находятся объекты, содержащие только соответствующие идентификаторы, введенные пользователем. Стоило их дозапрашивать на уровне сервиса и конструировать полноценный объект?

MAMKUH6OPIIIUK committed:
Выполнил домашнее задание №5

vlad-iv commented on spring-hw-05/src/main/java/ru/otus/spring/homework/oke/dao/AuthorsDaoJdbc.java:
```diff
@@ -0,0 +1,78 @@
+        MapSqlParameterSource parameterSource = new MapSqlParameterSource(parameters);
+        GeneratedKeyHolder keyHolder = new GeneratedKeyHolder();
+        jdbc.update(sql, parameterSource, keyHolder);
+        return new Author(keyHolder.getKey().longValue(), author.getName(), author.getMiddleName(),
```
можно вернуть author с установленным keyHolder.getKey().longValue()

MAMKUH6OPIIIUK replied:
Переделал модель - теперь без final полей, и значения можно устанавливать setter'ами. Ну и, соответственно, поменял в dao

vlad-iv commented on spring-hw-05/src/main/java/ru/otus/spring/homework/oke/dao/AuthorsDaoJdbc.java:
```diff
@@ -0,0 +1,78 @@
+        Map<String, Object> parameters = Collections.singletonMap("id", id);
+        try {
+            jdbc.update(sql, parameters);
+        } catch (DataIntegrityViolationException e) {
```
Предлагаю перенести обработку исключение в сервис

MAMKUH6OPIIIUK replied:
Не совсем согласен (или не совсем понял замечание), но перенёс в сервис.. Теперь он знает про исключения spring jdbc, специфичные для jdbc дао. 
Возможно, стоило добавлять предварительные дополнительные запросы (aka эмулировать работу constraint на стороне БД сервисом)?

vlad-iv commented on spring-hw-05/src/main/java/ru/otus/spring/homework/oke/dao/BooksDaoJdbc.java:
```diff
@@ -0,0 +1,294 @@
+        GeneratedKeyHolder keyHolder = new GeneratedKeyHolder();
+        try {
+            jdbc.update(sql, parameterSource, keyHolder);
+        } catch (DataIntegrityViolationException e) {
```
Предлагаю перенести из DAO обработку исключений, на слой сервиса или выше

MAMKUH6OPIIIUK replied:
Готово, перенёс

vlad-iv commented on spring-hw-05/src/main/java/ru/otus/spring/homework/oke/dao/GenresDaoJdbc.java:
```diff
@@ -0,0 +1,87 @@
+
+@Repository
+@RequiredArgsConstructor
+public class GenresDaoJdbc implements GenresDao {
```
Предлагаю придерживаться подхода, когда в префиксе указывается вариант реализации
вместо GenresDaoJdbc - > JdbcGenresDao

MAMKUH6OPIIIUK replied:
Готово, спасибо! Буду знать

vlad-iv commented on spring-hw-05/src/main/java/ru/otus/spring/homework/oke/exceptions/GenreNotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+package ru.otus.spring.homework.oke.exceptions;
+
+public class GenreNotFoundException extends RuntimeException {
```
Можно объединить все *NotFoundException в один и использовать только NotFoundException

MAMKUH6OPIIIUK replied:
Спасибо! Да, так попроще 👍 Объединил

vlad-iv commented on spring-hw-05/src/main/java/ru/otus/spring/homework/oke/service/BooksService.java:
```diff
@@ -0,0 +1,17 @@
+import java.util.List;
+
+public interface BooksService {
+    Book create(String title, String description, long authorId, long[] genreIds);
```
Предлагаю добавить Dto
так что бы в слой сервиса приходили Dto
а сервис их преобразовывал в модель , совершал логику, обращался к репозиторию
и возвращал Dto
Возможные подходы:
- Book create(BookDto)
или BookDto create(BookDto)
или BookResponseDto create(BookRequestDto)
или BookFullDto create(BookDto)

Предлагаю выбрать один из них и реализоваться для слоя сервис

MAMKUH6OPIIIUK replied:
Спасибо! Переделал. Сделал Dto для всего по схеме `BookResponseDto create(BookRequestDto)`. В ряде случаев dto полностью повторяют модель.
Также переделал модель книги в соответствии с рекомендацией в чате ДЗ - теперь в ней просто идентификаторы автора и жанров.  Конкретные объекты автора и жанров дозапрашиваются в сервисе. С одной стороны, из-за этого запросы стали чуть менее оптимальными (т.к. автор для dto запрашивается отдельно). С другой стороны, выглядит вроде логичнее и независимо друг от друга на слое дао.

vlad-iv commented on spring-hw-05/src/main/java/ru/otus/spring/homework/oke/service/BooksServiceImpl.java:
```diff
@@ -0,0 +1,56 @@
+        for (long genreId : genreIds) {
+            genreIdsForBook.add(new Genre(genreId, null));
+        }
+        Author bookAuthor = new Author(authorId, null, null, null, null);
```
Предлагаю добавить конструкторы с id и конструктор без id
так же для других сущностей

MAMKUH6OPIIIUK replied:
Готово, добавил 👍

vlad-iv commented on spring-hw-05/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,57 @@
+DROP TABLE IF EXISTS books_genres;
```
Предлагаю удалить DROP TABLE
Схема накатывается только на пустую БД

MAMKUH6OPIIIUK replied:
Да, спасибо, убрал

vlad-iv commented on spring-hw-05/src/test/resources/data.sql:
```diff
@@ -0,0 +1,26 @@
+insert into authors(name, middle_name, patronymic, surname) values ('Имя1', 'Среднее имя1', null, 'Фамилия1');
```
Так как ниже скрипт использует идентификаторы, предлагаю указывать их во всех insert

MAMKUH6OPIIIUK replied:
Добавил. Не добавил изначально потому, что при явном указании id в процессе insert не сдвигались значения autoincrement..  Теперь в data лежит ещё `alter table <name> alter column id restart with <max id + 1>;`

MAMKUH6OPIIIUK committed:
Исправления по результатам первой итерации проверки ДЗ№5

vlad-iv commented on spring-hw-05/src/main/java/ru/otus/spring/homework/oke/dao/BooksDao.java:
```diff
@@ -0,0 +1,18 @@
+
+    Book create(Book book);
+
+    Book update(Book book);
```
(можно лучше) void update(Book book);

MAMKUH6OPIIIUK replied:
Переделал и в dao, и в сервисе. В CLI теперь отображается только факт выполнения операции (успешно\не успешно)

vlad-iv commented on spring-hw-05/src/main/java/ru/otus/spring/homework/oke/dao/JdbcBooksDao.java:
```diff
@@ -0,0 +1,236 @@
+     * @param book книга для создания
+     * @return созданная книга с установленным идентификатором
+     */
+    @Transactional
```
Предлагаю все аннотации Transactional перенести из слоя репозиторий (JDBC)
в слой сервисов.
Причина, нам часто нужны не атомарные действия (из репозитория), а несколько действий сделанных одновременно (некоторая логика) (из сервиса, где объединяются разные действия вместе)

MAMKUH6OPIIIUK replied:
Спасибо! Перенёс в сервис

vlad-iv commented on spring-hw-05/src/main/java/ru/otus/spring/homework/oke/service/AuthorsServiceImpl.java:
```diff
@@ -0,0 +1,58 @@
+    private final AuthorMapper authorMapper;
+
+    @Override
+    public AuthorResponseDto create(AuthorRequestDto authorRequestDto) {
```
Предлагаю для всех публичных методов слоя сервис
добавить аннотацию Transactional 
а для тех методов, которые читают с вариантом readOnly

MAMKUH6OPIIIUK replied:
Готово

vlad-iv commented on spring-hw-05/src/main/java/ru/otus/spring/homework/oke/service/BooksServiceImpl.java:
```diff
@@ -0,0 +1,122 @@
+    public BookResponseDto update(long id, BookRequestDto bookRequestDto) {
+        try {
+            Book bookForUpdate = bookMapper.mapToBook(id, bookRequestDto);
+            Book updatedBook = this.booksDao.update(bookForUpdate);
```
Предлагаю добавить проверку (валидацию)
1. вызывать findById - проверить что книг есть
2. вызвать findById (или findByIds) для авторов и жанров - проверить что они тоже есть.
если что то не найдено выбрасываем NotFoundException из слоя сервис (не из репозитория!)

MAMKUH6OPIIIUK replied:
Готово. Добавил предварительные запросы автора и жанров при создании книги, а при обновлении - ещё и запрос старой версии книги. На основе этих данных теперь строится финальный dto. 
Прямо совсем избавиться от обработки исключений из dao не удалось, т.к. при поиске по конкретному id кидается исключение jdbc. Но при обработке на уровне сервиса заменил проверку конкретного исключения jdbc на обезличенное Exception. 
Возможно, стоило переписать dao, чтоб из них вообще ничего не кидалось и приходили какие-нибудь optional?

vlad-iv commented on spring-hw-05/src/main/java/ru/otus/spring/homework/oke/service/BooksServiceImpl.java:
```diff
@@ -0,0 +1,122 @@
+        List<Book> foundBooks = this.booksDao.findAll();
+        if (foundBooks.isEmpty()) {
+            return new ArrayList<>();
+        } else {
```
можно убрать else

MAMKUH6OPIIIUK replied:
Убрал, спасибо :-)

vlad-iv commented on spring-hw-05/src/main/java/ru/otus/spring/homework/oke/service/BooksServiceImpl.java:
```diff
@@ -0,0 +1,122 @@
+    public List<BookResponseDto> findAll() {
+        List<Book> foundBooks = this.booksDao.findAll();
+        if (foundBooks.isEmpty()) {
+            return new ArrayList<>();
```
Лучше Collection emptyList

MAMKUH6OPIIIUK replied:
Заменил на Collections.EMPTY_LIST

MAMKUH6OPIIIUK committed:
Исправления по результатам второй итерации проверки ДЗ№5

1. Методы update у дао и сервиса для работы с книгами больше не возвращают объекты. В CLI пишется только о факте успешного завершения обновления
2. Управление транзакциями перенесено из дао в сервисы
3. На уровне сервисов добавлена валидация:
3.1. BookService теперь предварительно проверяет существование автора и жанров, которые необходимо назначить книге при её создании или обновлении. При обновлении предварительно проверяется и существование книги. При отсутствии кидается исключение
3.2. AuthorsService перед удалением автора теперь запрашивает его книги. При наличии кидается исключение.
4. На уровне дао добавились новые вспомогательные методы для реализации валидации в сервисах:
4.1. GenresDao - метод поиска жанров по набору идентификаторов (запрос с "id IN ()"
4.2. BooksDao - метод поиска всех книг по идентификатору автора.
5. Тесты дополнены проверками для новой функциональности.
